CMakeLists.txt

```diff
@@ -0,0 +1,356 @@
+cmake_minimum_required(VERSION 3.12)
+
+project(coap VERSION 4.3.0 LANGUAGES CXX C)
+
+set(LIBCOAP_API_VERSION 2)
+
+#
+# options to tweak the library
+#
+
+option(ENABLE_DTLS "Enable building with DTLS support" ON)
+option(WITH_OPENSSL "Use OpenSSL for DTLS functions" ON)
+option(WITH_GNUTLS "Use GnuTLS for DTLS functions" OFF)
+option(WITH_TINYDTLS "Use TinyDTLS for DTLS functions" OFF)
+option(WITH_EPOLL "compile with epoll support" OFF)
+option(ENABLE_SMALL_STACK "Define if the system has small stack size" OFF)
+option(ENABLE_TESTS "build also tests" OFF)
```

Comment:
With ENABLE_TESTS=ON, WITH_OPENSSL=ON testdriver fails to build, but WITH_GNTTLS=ON this works.

Reply:
what is the error? 

works for me on ubunt 18.04, gcc 9, cmake 3.13
```bash
rm -rf _build
mkdir _build
cd _build
cmake .. -DENABLE_TESTS=ON -DWITH_OPENSSL=ON
cmake --build . -j
```

Reply:
Ubuntu 18.04, gcc 7.4.0, cmake 3.10.2
````
[ 62%] Building C object CMakeFiles/coap.dir/src/coap_openssl.c.o
[ 65%] Linking C static library libcoap.a
[ 65%] Built target coap
Scanning dependencies of target testdriver
[ 68%] Building C object CMakeFiles/testdriver.dir/tests/testdriver.c.o
[ 71%] Building C object CMakeFiles/testdriver.dir/tests/test_encode.c.o
[ 75%] Building C object CMakeFiles/testdriver.dir/tests/test_error_response.c.o
[ 78%] Building C object CMakeFiles/testdriver.dir/tests/test_options.c.o
[ 81%] Building C object CMakeFiles/testdriver.dir/tests/test_pdu.c.o
[ 84%] Building C object CMakeFiles/testdriver.dir/tests/test_sendqueue.c.o
[ 87%] Building C object CMakeFiles/testdriver.dir/tests/test_session.c.o
[ 90%] Building C object CMakeFiles/testdriver.dir/tests/test_tls.c.o
[ 93%] Building C object CMakeFiles/testdriver.dir/tests/test_uri.c.o
[ 96%] Building C object CMakeFiles/testdriver.dir/tests/test_wellknown.c.o
[100%] Linking C executable testdriver
CMakeFiles/testdriver.dir/tests/test_tls.c.o: In function `t_tls1':
test_tls.c:(.text+0x17): undefined reference to `coap_dtls_is_supported'
CMakeFiles/testdriver.dir/tests/test_tls.c.o: In function `t_tls2':
test_tls.c:(.text+0x66): undefined reference to `coap_get_tls_library_version'
...
````
Works fine with -DWITH_GNUTLS=ON -DWITH_OPENSSL=OFF
````
[ 62%] Building C object CMakeFiles/coap.dir/src/coap_gnutls.c.o
[ 65%] Linking C static library libcoap.a
[ 65%] Built target coap
Scanning dependencies of target testdriver
[ 68%] Building C object CMakeFiles/testdriver.dir/tests/testdriver.c.o
[ 71%] Building C object CMakeFiles/testdriver.dir/tests/test_encode.c.o
[ 75%] Building C object CMakeFiles/testdriver.dir/tests/test_error_response.c.o
[ 78%] Building C object CMakeFiles/testdriver.dir/tests/test_options.c.o
[ 81%] Building C object CMakeFiles/testdriver.dir/tests/test_pdu.c.o
[ 84%] Building C object CMakeFiles/testdriver.dir/tests/test_sendqueue.c.o
[ 87%] Building C object CMakeFiles/testdriver.dir/tests/test_session.c.o
[ 90%] Building C object CMakeFiles/testdriver.dir/tests/test_tls.c.o
[ 93%] Building C object CMakeFiles/testdriver.dir/tests/test_uri.c.o
[ 96%] Building C object CMakeFiles/testdriver.dir/tests/test_wellknown.c.o
[100%] Linking C executable testdriver
[100%] Built target testdriver
````

Reply:
Issue found - there was a `include/coap2/coap_config.h` and `_build/include/coap2/coap_config.h` .  The former file had a `#define HAVE_GNUTLS "1"` and the latter `#define HAVE_OPENSSL "1"`.

I am not sure how or when `include/coap2/coap_config.h` was created, but someone else will do the same thing. There also will be an issue if `include/coap2/coap.h` exists.

Certainly target_include_directories() is building the include directory search order incorrectly
````
target_include_directories(
  ${COAP_LIBRARY_NAME}
  PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include/coap2>
         # some *.c are still including the filename instead of
         # "coap2/header.h"
         $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include/>
         # config headers are generated during configuration time
         $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/>
         $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/coap2/>
         $<INSTALL_INTERFACE:include/>
         $<$<BOOL:${HAVE_LIBGNUTLS}>:${GNUTLS_INCLUDE_DIR}>)  
````
with CMAKE_CURRENT_LIST_DIR being searched before CMAKE_CURRENT_BINARY_DIR.  However, the compiler, even with CMAKE_CURRENT_BINARY_DIR having a higher priority (which is a correct and needed fix) still chooses the non _build version because the first #include header in src/*.c is "coap_internal.h" and so when "coap_config.h" is looked up as the first #include in "coap_internal.h" and finds "coap_config.h" in the same directory, it uses it.  Hmm.

Easy to test by just creating `include/coap2/coap_config.h` in the same directory as `include/coap2/coap_config.h.in` with the following definition and running the cmake sequence.
````
#ifndef COAP_CONFIG_H_
#define COAP_CONFIG_H_
#bad
#endif /* COAP_CONFIG_H_ */
````

The only way around this appears to be the replacing of #include `"coap_internal.h"` with
````
#include "coap_config.h"
#include "coap_internal.h"
````
in all the appropriate src/*.c files if someone is not going to get stung in the future.

Reply:
`include/coap2/coap_config.h` is in the git ignore list, probably you tried first autoconf which polutes the whole source directory...